Build two keras deep learning models on the fashion MNIST data with the following requirements (run for epochs = 20, and keep the validation set the exact same way it was in the code we did in class):

- There should be two (dense) hidden layers of size 50 and 20 with the relu activation function. Keep the output layer the way it was (with softmax and ten nodes).
- In the second model do the same, but switch the order of the layers. That is, the first hidden layer should have 20 nodes and the second one should have 50 nodes.
- As you know when you run “model.fit” you should see the results for the 20 epochs. Based on the accuracy numbers that are printed at the end of the 20 epochs, which of the two networks appear to be better?
- ONLY KEEP THE CODE THIS IS NEEDED – there is no need to print data, or an image, or model.layers or any of the other things that was there in the class code. Make sure the code you have is exactly what was needed to answer the parts above.

In [2]:
import tensorflow as tf
from tensorflow import keras
import matplotlib
import matplotlib.pyplot as plt

In [3]:
fashion_mnist = keras.datasets.fashion_mnist

(X_train_full, y_train_full), (X_test, y_test) = fashion_mnist.load_data()

In [4]:
#create validation data as well, plus scaling down x values from 1-255 to 0-1 scale
X_valid, X_train = X_train_full[:5000] / 255.0, X_train_full[5000:] / 255.0
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

In [5]:
class_names = ["T-shirt/top", "Trouser", "Pullover", "Dress", "Coat",
               "Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"]
class_names[y_train[0]]

'Coat'

In [6]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(50, activation="relu"))
model.add(keras.layers.Dense(20, activation="relu"))
model.add(keras.layers.Dense(10, activation="softmax"))

In [7]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [8]:
history = model.fit(X_train, y_train, epochs=20,validation_data=(X_valid, y_valid))

Epoch 1/20
1719/1719 [==============================] - 2s 875us/step - loss: 0.8784 - accuracy: 0.7041 - val_loss: 0.5631 - val_accuracy: 0.8148
Epoch 2/20
1719/1719 [==============================] - 1s 684us/step - loss: 0.5294 - accuracy: 0.8176 - val_loss: 0.4916 - val_accuracy: 0.8328
Epoch 3/20
1719/1719 [==============================] - 1s 698us/step - loss: 0.4753 - accuracy: 0.8340 - val_loss: 0.4391 - val_accuracy: 0.8500
Epoch 4/20
1719/1719 [==============================] - 1s 684us/step - loss: 0.4464 - accuracy: 0.8436 - val_loss: 0.4317 - val_accuracy: 0.8540
Epoch 5/20
1719/1719 [==============================] - 1s 679us/step - loss: 0.4268 - accuracy: 0.8492 - val_loss: 0.4494 - val_accuracy: 0.8408
Epoch 6/20
1719/1719 [==============================] - 1s 678us/step - loss: 0.4103 - accuracy: 0.8566 - val_loss: 0.3970 - val_accuracy: 0.8648
Epoch 7/20
1719/1719 [==============================] - 1s 661us/step - loss: 0.3970 - accuracy: 0.8607 - val_loss: 0.3986 -

In [9]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 0s 894us/step - loss: 91.9494 - accuracy: 0.7955


[91.94937896728516, 0.7954999804496765]

The accuracy is 79%

In [11]:
model2 = keras.models.Sequential()
model2.add(keras.layers.Flatten(input_shape=[28, 28]))
model2.add(keras.layers.Dense(20, activation="relu"))
model2.add(keras.layers.Dense(50, activation="relu"))
model2.add(keras.layers.Dense(10, activation="softmax"))

In [12]:
model2.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [13]:
history = model2.fit(X_train, y_train, epochs=20,validation_data=(X_valid, y_valid))

Epoch 1/20
1719/1719 [==============================] - 1s 660us/step - loss: 0.9057 - accuracy: 0.6895 - val_loss: 0.5963 - val_accuracy: 0.7858
Epoch 2/20
1719/1719 [==============================] - 1s 600us/step - loss: 0.5625 - accuracy: 0.8023 - val_loss: 0.5083 - val_accuracy: 0.8204
Epoch 3/20
1719/1719 [==============================] - 1s 636us/step - loss: 0.5019 - accuracy: 0.8233 - val_loss: 0.4634 - val_accuracy: 0.8390
Epoch 4/20
1719/1719 [==============================] - 1s 634us/step - loss: 0.4712 - accuracy: 0.8338 - val_loss: 0.4526 - val_accuracy: 0.8456
Epoch 5/20
1719/1719 [==============================] - 1s 614us/step - loss: 0.4502 - accuracy: 0.8416 - val_loss: 0.4370 - val_accuracy: 0.8510
Epoch 6/20
1719/1719 [==============================] - 1s 640us/step - loss: 0.4329 - accuracy: 0.8480 - val_loss: 0.4313 - val_accuracy: 0.8484
Epoch 7/20
1719/1719 [==============================] - 1s 625us/step - loss: 0.4183 - accuracy: 0.8515 - val_loss: 0.4129 -

The accuracy for the second model is 86%

We can say that the second model is better than the first model as the accuracy for the first model is 79 which less than 89%.